### PR TITLE
Query Module: Logical plan Pt.

### PR DESCRIPTION
**Pull request by lujiajing1126**
This PR focuses on the logical plan of the query module.

## Logical Plan

The logical plan is generated from QueryCriteria and is organized as a tree.

### Resolved and Unresolved plan/expr

We consider the `QueryCriteria` in an unresolved status (i.e. it does not contain any metadata of the entities), and directly generate an **Unresolved** logical plan first and further resolve it with metadata.

For example, 

```go
// fieldRef is the reference to the field
// also it holds the definition/schema of the field
type fieldRef struct {
	// name defines the key of the field
	name string
	// spec contains the index of the key in the schema, as well as the underlying FieldSpec
	spec *fieldSpec
}

func NewFieldRef(fieldName string) *fieldRef {
	return &fieldRef{
		name: fieldName,
	}
}
```

when we create the `fieldRef`, it stays in the `Unresolved` status which mean it does not know anything about the typing, even cannot guarantee a field with the given name exists in the definition.

The separation of `Resolved` and `Unresolved` statuses would make it possible to build a logical plan with DSL much easier.

```go
correctPlan, err := logical.Projection(logical.Limit(
		logical.Offset(logical.OrderBy(
			logical.Selection(
				logical.Scan(uint64(sT.UnixNano()), uint64(eT.UnixNano()), traceMetadata, "service_id", "trace_id", "service_instance_id", "http.method", "service_name"),
				logical.Eq(logical.NewFieldRef("http.method"), logical.Str("GET")),
				logical.Eq(logical.NewFieldRef("service_name"), logical.Str("my_app")),
			), "service_instance_id", apiv1.SortDESC,
		), 10), 5,
	), []string{"service_id", "trace_id"}).Analyze(schema)
```

This can be used to compare with the generated plan, which would significantly improve the testibility. For equality test, we also introduce `T Equal(T) bool` like method for Plan, Expr, and Metadata as suggested in [`go-cmp` document](https://pkg.go.dev/github.com/google/go-cmp/cmp#Equal).

### Inspection

The logical plan can be printed as string, 

```
+----Limit: 5
     +----Offset: 10
          +----OrderBy: #service_instance_id<String>, sort=DESC
               +----Selection: [#http.method<String> EQ GET AND #service_name<String> EQ my_app]
                    +----Scan: startTime=1624185419600386000,endTime=1624196219600388000,Metadata{group=default,name=trace}; projection=#service_id<String>, #trace_id<String>, #service_instance_id<String>, #http.method<String>, #service_name<String>
```

## Optimizer

Removed.

### CODE REVIEW

File: pkg/logical/optimizer_test.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+	plan, err = logical.NewProjectionPushDown().Apply(plan)
+	assert.NoError(err)
+	assert.NotNil(plan)
+	fmt.Print(logical.Format(plan))
+
+	correctPlan, err := logical.Projection(logical.Limit(
+		logical.Offset(logical.OrderBy(
+			logical.Selection(
+				logical.Scan(uint64(sT.UnixNano()), uint64(eT.UnixNano()), traceMetadata, "service_id", "trace_id", "service_instance_id", "http.method", "service_name"),
```
Comment (hanahmily) [inline pkg/logical/optimizer_test.go:73]: Why pushing fields into the projection list since the input doesn't indicate they will be returned.

**Review by hanahmily — CHANGES_REQUESTED** [review]
From your screenshots, the `Scan` is the first operation even fields exist. From the design, `Select` which should be the first one parsed ChunkIDs from `index`, then the results will be passed to  `FetchEntity`(I didn't see this step in the parsed plan). The `Scan` will be used only there's no any `Selection` input, which means it will be picked rarely.

**Comment by lujiajing1126** [comment]
> From your screenshots, the `Scan` is the first operation even fields exist. From the design, `Select` which should be the first one parsed ChunkIDs from `index`, then the results will be passed to `FetchEntity`(I didn't see this step in the parsed plan). The `Scan` will be used only there's no any `Selection` input, which means it will be picked rarely.

As I understood, the logical plan does not care about the indexes, it just prepares the metadata and resolves fields so that the existence of these fields which are referenced can be guaranteed.

As we've discussed in the last PR, we can select the indexes while generating physical plans based on cost-first consideration.

**Comment by hanahmily** [comment]
> > From your screenshots, the `Scan` is the first operation even fields exist. From the design, `Select` which should be the first one parsed ChunkIDs from `index`, then the results will be passed to `FetchEntity`(I didn't see this step in the parsed plan). The `Scan` will be used only there's no any `Selection` input, which means it will be picked rarely.
> 
> As I understood, the logical plan does not care about the indexes, it just prepares the metadata and resolves fields so that the existence of these fields which are referenced can be guaranteed.
> 
> As we've discussed in the last PR, we can select the indexes while generating physical plans based on cost-first consideration.

Not that. cost-based optimization only takes place when a field belongs to more than one index. Supposing `service_id` is indexed by `service_id` + `instance_id` and `service_id` + `endpoint_id`, we should leverage cost-based statistics to determine which one should be used. In this case, the combination of `service_id` and `instance_id` has fewer cardinalities, which will get less cost. Based on that, the query physical optimizer should pick up it instead of `service_id` + `endpoint_id`.

As I mentioned in another PR, we don't support the composite index for now. We don't have to implement the above plan optimization. 

>  the logical plan does not care about the indexes

That's the convention of a traditional SQL database due to the complex index type, for example, single, composite, unique, group by and etc. Such a database can't determine query paths without statistics. But the query of BanyanDB doesn't, which knows how to access single fields index based on the criteria. 

We also remove `tags` from query criteria, which's due to the huge cost of `Scan` filtering. As I mentioned, the query plan picks up `Scan` when no fields are input.

File: pkg/logical/analyzer.go [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+					} else {
+						return nil, InvalidConditionTypeErr
+					}
+				} else {
+					return nil, TypePairInspectionErr
+				}
+			} else {
+				return nil, PairQueryInspectionErr
+			}
+		}
```
[the remark this answers is not in the document]
Comment (lujiajing1126) [reply pkg/logical/analyzer.go:113]: Fixed

File: pkg/logical/analyzer_test.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+		Spec:        *criteria.Metadata(nil),
+	}
+
+	schema, err := ana.BuildTraceSchema(context.TODO(), *metadata)
+	assert.NoError(err)
+
+	plan, err := ana.Analyze(context.TODO(), criteria, metadata, schema)
+	assert.NoError(err)
+	assert.NotNil(plan)
+	fmt.Print(logical.Format(plan))
```
[the remark this answers is not in the document]
Comment (lujiajing1126) [reply pkg/logical/analyzer_test.go:58]: Now I use DSL to produce a correct plan to compare with the generated one

File: pkg/logical/analyzer_test.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+}
+
+func TestAnalyzer_Fields_FieldNotDefined(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (lujiajing1126) [reply pkg/logical/analyzer_test.go:91]: Only the field is not define in this case

File: pkg/logical/optimizer_test.go (hunk 1 above)
Comment (lujiajing1126) [reply pkg/logical/optimizer_test.go:73]: The optimizer is removed since it is not necessary now

**Comment by lujiajing1126** [comment]
`TraceIDFetch` and `IndexScan` have been introduced as we've discussed last time.

@hanahmily Pls have a review again, thx

File: pkg/logical/analyzer_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (hanahmily) [reply pkg/logical/analyzer_test.go]: Cool!

**Review by hanahmily — APPROVED** [review]
LGTM.

**Pull request merged**